File: swift-bridge/Sources/AVAudioBridge/EnvironmentNode.swift
import AVFoundation
import Foundation

struct ListenerPositionPayload: Codable {
    let x: Float
    let y: Float
    let z: Float
}

struct ListenerOrientationPayload: Codable {
    let yaw: Float
    let pitch: Float
    let roll: Float
}

struct DistanceAttenuationPayload: Codable {
    let model: Int32
    let referenceDistance: Float
    let maximumDistance: Float
    let rolloffFactor: Float
}

@_cdecl("av_audio_environment_node_create")
public func av_audio_environment_node_create() -> UnsafeMutableRawPointer? {
    Unmanaged.passRetained(AVAudioEnvironmentNode()).toOpaque()
}

@_cdecl("av_audio_environment_node_release")
public func av_audio_environment_node_release(_ ptr: UnsafeMutableRawPointer?) {
    guard let ptr else { return }
    Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).release()
}

@_cdecl("av_audio_environment_node_set_listener_position")
public func av_audio_environment_node_set_listener_position(
    _ ptr: UnsafeMutableRawPointer,
    _ x: Float,
    _ y: Float,
    _ z: Float
) {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    node.listenerPosition = AVAudio3DPoint(x: x, y: y, z: z)
}

@_cdecl("av_audio_environment_node_get_listener_position_json")
public func av_audio_environment_node_get_listener_position_json(
    _ ptr: UnsafeMutableRawPointer,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    let payload = ListenerPositionPayload(
        x: node.listenerPosition.x,
        y: node.listenerPosition.y,
        z: node.listenerPosition.z
    )
    do {
        return ffiString(try avaEncodeJSON(payload))
    } catch {
        outError?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

@_cdecl("av_audio_environment_node_set_listener_orientation")
public func av_audio_environment_node_set_listener_orientation(
    _ ptr: UnsafeMutableRawPointer,
    _ yaw: Float,
    _ pitch: Float,
    _ roll: Float
) {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    node.listenerAngularOrientation = AVAudio3DAngularOrientation(yaw: yaw, pitch: pitch, roll: roll)
}

@_cdecl("av_audio_environment_node_get_listener_orientation_json")
public func av_audio_environment_node_get_listener_orientation_json(
    _ ptr: UnsafeMutableRawPointer,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    let orientation = node.listenerAngularOrientation
    let payload = ListenerOrientationPayload(yaw: orientation.yaw, pitch: orientation.pitch, roll: orientation.roll)
    do {
        return ffiString(try avaEncodeJSON(payload))
    } catch {
        outError?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

@_cdecl("av_audio_environment_node_set_distance_attenuation")
public func av_audio_environment_node_set_distance_attenuation(
    _ ptr: UnsafeMutableRawPointer,
    _ model: Int32,
    _ referenceDistance: Float,
    _ maximumDistance: Float,
    _ rolloffFactor: Float
) {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    let params = node.distanceAttenuationParameters
    params.distanceAttenuationModel = AVAudioEnvironmentDistanceAttenuationModel(rawValue: Int(model)) ?? .exponential
    params.referenceDistance = referenceDistance
    params.maximumDistance = maximumDistance
    params.rolloffFactor = rolloffFactor
}

@_cdecl("av_audio_environment_node_get_distance_attenuation_json")
public func av_audio_environment_node_get_distance_attenuation_json(
    _ ptr: UnsafeMutableRawPointer,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    let params = node.distanceAttenuationParameters
    let payload = DistanceAttenuationPayload(
        model: Int32(params.distanceAttenuationModel.rawValue),
        referenceDistance: params.referenceDistance,
        maximumDistance: params.maximumDistance,
        rolloffFactor: params.rolloffFactor
    )
    do {
        return ffiString(try avaEncodeJSON(payload))
    } catch {
        outError?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

@_cdecl("av_audio_environment_node_set_reverb_blend")
public func av_audio_environment_node_set_reverb_blend(
    _ ptr: UnsafeMutableRawPointer,
    _ blend: Float
) {
    let node = Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue()
    node.reverbParameters.enable = blend > 0
    node.reverbParameters.level = blend
}

@_cdecl("av_audio_environment_node_get_reverb_blend")
public func av_audio_environment_node_get_reverb_blend(_ ptr: UnsafeMutableRawPointer) -> Float {
    Unmanaged<AVAudioEnvironmentNode>.fromOpaque(ptr).takeUnretainedValue().reverbParameters.level
}

File: src/environment_node.rs
//! [`AudioEnvironmentNode`] — 3D spatial audio support.

#![allow(
    clippy::missing_errors_doc,
    clippy::must_use_candidate,
    clippy::module_name_repetitions
)]

use core::ffi::{c_char, c_void};
use core::ptr;

use serde::Deserialize;

use crate::error::{from_swift, AVAudioError};
use crate::ffi;
use crate::node::AudioNodeHandle;
use crate::util::parse_json_and_free;

/// Listener position in 3D space.
#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AudioListenerPosition {
    /// X coordinate.
    pub x: f32,
    /// Y coordinate.
    pub y: f32,
    /// Z coordinate.
    pub z: f32,
}

/// Listener orientation in angular degrees.
#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AudioListenerOrientation {
    /// Yaw angle.
    pub yaw: f32,
    /// Pitch angle.
    pub pitch: f32,
    /// Roll angle.
    pub roll: f32,
}

/// Distance attenuation parameters.
#[derive(Debug, Clone, Copy, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AudioDistanceAttenuation {
    /// Raw `AVAudioEnvironmentDistanceAttenuationModel` value.
    pub model: i32,
    /// Reference distance in meters.
    pub reference_distance: f32,
    /// Maximum distance in meters.
    pub maximum_distance: f32,
    /// Rolloff factor.
    pub rolloff_factor: f32,
}

/// Wraps an `AVAudioEnvironmentNode`.
pub struct AudioEnvironmentNode {
    pub(crate) ptr: *mut c_void,
}

unsafe impl Send for AudioEnvironmentNode {}

impl Drop for AudioEnvironmentNode {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::av_audio_environment_node_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

impl AudioNodeHandle for AudioEnvironmentNode {
    fn as_node_ptr(&self) -> *mut c_void {
        self.ptr
    }
}

impl AudioEnvironmentNode {
    /// Creates a new environment node.
    pub fn new() -> Result<Self, AVAudioError> {
        let ptr = unsafe { ffi::av_audio_environment_node_create() };
        if ptr.is_null() {
            return Err(AVAudioError::OperationFailed(
                "failed to create AVAudioEnvironmentNode".into(),
            ));
        }
        Ok(Self { ptr })
    }

    /// Sets the listener position.
    pub fn set_listener_position(&self, x: f32, y: f32, z: f32) {
        unsafe { ffi::av_audio_environment_node_set_listener_position(self.ptr, x, y, z) };
    }

    /// Returns the listener position.
    pub fn listener_position(&self) -> Result<AudioListenerPosition, AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe {
            ffi::av_audio_environment_node_get_listener_position_json(self.ptr, &mut err)
        };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    /// Sets the listener orientation.
    pub fn set_listener_orientation(&self, yaw: f32, pitch: f32, roll: f32) {
        unsafe {
            ffi::av_audio_environment_node_set_listener_orientation(self.ptr, yaw, pitch, roll);
        };
    }

    /// Returns the listener orientation.
    pub fn listener_orientation(&self) -> Result<AudioListenerOrientation, AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe {
            ffi::av_audio_environment_node_get_listener_orientation_json(self.ptr, &mut err)
        };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    /// Sets the distance attenuation parameters.
    pub fn set_distance_attenuation(
        &self,
        model: i32,
        reference_distance: f32,
        maximum_distance: f32,
        rolloff_factor: f32,
    ) {
        unsafe {
            ffi::av_audio_environment_node_set_distance_attenuation(
                self.ptr,
                model,
                reference_distance,
                maximum_distance,
                rolloff_factor,
            );
        };
    }

    /// Returns the distance attenuation parameters.
    pub fn distance_attenuation(&self) -> Result<AudioDistanceAttenuation, AVAudioError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe {
            ffi::av_audio_environment_node_get_distance_attenuation_json(self.ptr, &mut err)
        };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    /// Sets the environment reverb blend level.
    pub fn set_reverb_blend(&self, blend: f32) {
        unsafe { ffi::av_audio_environment_node_set_reverb_blend(self.ptr, blend) };
    }

    /// Returns the environment reverb blend level.
    pub fn reverb_blend(&self) -> f32 {
        unsafe { ffi::av_audio_environment_node_get_reverb_blend(self.ptr) }
    }
}
